pkg/process_manager.go

```diff
@@ -18,46 +26,335 @@ const (
+		func() {
+			s.waiter.L.Lock()
+			defer s.waiter.L.Unlock()
+			s.waiter.Wait()
+		}()
+	}
+
+	// Read historic or new data.
+	n, err = s.buf.Read(p)
+
+	if err == io.EOF && !s.eof.Load() {
+		// Don't return EOF until the stream has closed.
+		err = nil
+
+		// Since the end of the buffer has been reached at least once, the reads
+		// are now caught up and future reads should wait for the waiter.
+		s.caughtUp = true
+
+		// Try again to wait for more data or EOF.
+		n, err = s.Read(p)
+	}
+
+	return
+}
+
+type process struct {
+	cancel         context.CancelFunc
+	statusLock     sync.RWMutex
+	status         ProcessStatus
+	terminated     atomic.Bool
+	terminatedCond *sync.Cond
+	stdout         processStream
+	stderr         processStream
+}
+
+func (s *process) markTerminated() {
+	// Notify any sleeping streams that there won't be any more output.
+	s.stdout.eof.Store(true)
+	s.stdout.waiter.Broadcast()
+	s.stderr.eof.Store(true)
+	s.stderr.waiter.Broadcast()
+
+	s.terminated.Store(true)
+	s.terminatedCond.Broadcast()
+}
+
+func (s *process) waitTerminated() {
+	if !s.terminated.Load() {
+		func() {
+			s.terminatedCond.L.Lock()
+			defer s.terminatedCond.L.Unlock()
+			s.terminatedCond.Wait()
+		}()
+	}
 }
 
 type ProcessManager struct {
-	// TODO
+	processesLock sync.RWMutex
+	processes     []*process
 }
 
-func (*ProcessManager) Start() uint32 {
-	panic("TODO")
+var ErrUnknownID = errors.New("unknown process ID")
+
+func (s *ProcessManager) acquireProcess(id int) (*process, error) {
+	s.processesLock.RLock()
+	defer s.processesLock.RUnlock()
+
+	if id >= len(s.processes) || id < 0 {
+		return &process{}, ErrUnknownID
+	}
+
+	return s.processes[id], nil
+}
+
+// Start starts a process and returns its ID.
+//
+// If `path` contains no path separators, the location is resolved from `$PATH`.
+// Note that `args` includes `argv[0]`.
+//
+// This function fails if starting the process fails.
+func (s *ProcessManager) Start(path string, args ...string) (int, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	cmd := exec.CommandContext(ctx, path)
+	cmd.Args = args
+
+	handleStream := func(pipe io.Reader, stream processStream) {
+		buf := make([]byte, 1024)
+		for {
+			n, err := pipe.Read(buf)
+
+			stream.buf.writer.Write(buf[:n])
+			stream.waiter.Broadcast()
+
+			if errors.Is(err, io.EOF) {
+				stream.eof.Store(true)
+				break
+			} else if err != nil {
+				// Ideally we'd communicate this error somehow (either sending
+				// it to the caller somehow or just logging it) but this is not
+				// done for simplicity. Instead, we'll just mark the stream as
+				// having reached EOF.
+				stream.eof.Store(true)
+				break
+			}
+		}
+	}
+
+	stdoutStream := processStream{
+		buf:    new(multiBufferWriter).newReader(),
+		waiter: sync.NewCond(new(sync.Mutex)),
+		eof:    new(atomic.Bool),
+	}
+	stdoutPipe, err := cmd.StdoutPipe()
+	if err != nil {
+		cancel()
+		return 0, err
+	}
+	go handleStream(stdoutPipe, stdoutStream)
```

Comment:
Your `multiBufferWriter` already implements `io.Writer`, so we shouldn't need to run our own goroutines to read data from a pipe and then write it to the writer.

We can instead set `Stdout` and `Stderr` on the `exec.Cmd` to an instance of your buffer and Go will take care of the goroutines to copy data. The added advantage of this (aside from extra code that you don't have to write or maintain) is the command's `Wait()` method will automatically handle closing the data copying goroutines and making sure they shut down gracefully before returning (which is a property your current implementation doesn't have, since you call `markTerminated` coordinating with `handleStream`).

Reply:
This was great advice. As a result I've refactored away `processStream` completely and now everything is handled by `multiBuffer{Writer,Reader}` but still in a general way. One thing that's slightly too bad is that `exec.Cmd`'s `Stdout` and `Stderr` fields are `io.Writer`s rather than `io.WriteCloser`s which would've made this even easier. Not a huge deal though, just means I need to manually `Close` them after `(*exec.Cmd).Wait`ing.